driver-sync/src/main/com/mongodb/client/internal/MongoChangeStreamCursorImpl.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.internal;
+
+import com.mongodb.client.MongoChangeStreamCursor;
+import com.mongodb.lang.Nullable;
+import com.mongodb.operation.BatchCursor;
+import org.bson.BsonDocument;
+
+public class MongoChangeStreamCursorImpl<T> extends MongoBatchCursorAdapter<T> implements MongoChangeStreamCursor<T> {
```

Comment:
Is this always really `MongoChangeStreamCursorImpl<ChangeStreamDocument<TResult>>` if so you could cache the current resume token and return that rather than the final one from the `batchCursor`

Reply:
Agreed, I think you have to, given that once this class knows how far the cursor has actually been iterated by the application.

Reply:
But it's not always a ChangeStreamDocument.   

I think to make this work we're going to have to create batch cursors of type BsonDocument, and do the conversion to `T` all the way up here.  Otherwise there is no way to grab the _id from each document as it's returned from next() or tryNext().

Reply:
I think if its not a `ChangeStreamDocument` it goes through a different code path.

Reply:
The conversion to `T` is now performed in the `nextInBatch()` method.